File: dotnet/CucumberExpressions.Tests/CucumberExpressionParserTest.java
package io.cucumber.cucumberexpressions;

import io.cucumber.cucumberexpressions.Ast.Node;
import org.junit.jupiter.api.extension.ParameterContext;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.converter.ArgumentConversionException;
import org.junit.jupiter.params.converter.ArgumentConverter;
import org.junit.jupiter.params.converter.ConvertWith;
import org.junit.jupiter.params.provider.MethodSource;
import org.yaml.snakeyaml.Yaml;

import java.io.IOException;
import java.io.InputStream;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;
import java.util.stream.Collectors;

import static java.nio.file.Files.newDirectoryStream;
import static java.nio.file.Files.newInputStream;
import static org.hamcrest.MatcherAssert.assertThat;
import static org.hamcrest.Matchers.is;
import static org.junit.jupiter.api.Assertions.assertThrows;

class CucumberExpressionParserTest {

    private final CucumberExpressionParser parser = new CucumberExpressionParser();

    private static List<Path> acceptance_tests_pass() throws IOException {
        List<Path> paths = new ArrayList<>();
        newDirectoryStream(Paths.get("..", "testdata", "cucumber-expression", "parser")).forEach(paths::add);
        paths.sort(Comparator.naturalOrder());
        return paths;
    }

    @ParameterizedTest
    @MethodSource
    void acceptance_tests_pass(@ConvertWith(Converter.class) Expectation expectation) {
        if (expectation.exception == null) {
            Node node = parser.parse(expectation.expression);
            assertThat(node, is(expectation.expected_ast.toNode()));
        } else {
            CucumberExpressionException exception = assertThrows(
                    CucumberExpressionException.class,
                    () -> parser.parse(expectation.expression));
            assertThat(exception.getMessage(), is(expectation.exception));
        }
    }

    static class Expectation {
        public String expression;
        public YamlableNode expected_ast;
        public String exception;
    }

    static class Converter implements ArgumentConverter {
        Yaml yaml = new Yaml();

        @Override
        public Expectation convert(Object source, ParameterContext context) throws ArgumentConversionException {
            try {
                Path path = (Path) source;
                InputStream inputStream = newInputStream(path);
                return yaml.loadAs(inputStream, Expectation.class);
            } catch (IOException e) {
                throw new ArgumentConversionException("Could not load " + source, e);
            }
        }
    }

    static class YamlableNode {
        public Ast.Node.Type type;
        public List<YamlableNode> nodes;
        public String token;
        public int start;
        public int end;

        public Node toNode() {
            if (token != null) {
                return new Node(type, start, end, token);
            } else {
                return new Node(type, start, end, nodes.stream().map(YamlableNode::toNode).collect(Collectors.toList()));
            }
        }
    }
}

File: dotnet/CucumberExpressions.Tests/CucumberExpressionParserTest.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using CucumberExpressions.Parsing;
using FluentAssertions;
using Xunit;
using Xunit.Abstractions;

namespace CucumberExpressions.Tests;

public class CucumberExpressionParserTest : TestBase
{
    private readonly ITestOutputHelper _testOutputHelper;
	private readonly CucumberExpressionParser _parser = new();

    public CucumberExpressionParserTest(ITestOutputHelper testOutputHelper)
    {
        _testOutputHelper = testOutputHelper;
    }

    public static IEnumerable<object[]> acceptance_tests_pass_data()
		=> GetTestDataFiles("cucumber-expression", "parser")
			.Select(file => new object[]
			{
				Path.GetFileNameWithoutExtension(file),
				ParseYaml<Expectation>(file)
			});

	[Theory, MemberData(nameof(acceptance_tests_pass_data))]
	[System.Diagnostics.CodeAnalysis.SuppressMessage("Usage", "xUnit1026:Theory methods should use all of their parameters", Justification = "<Pending>")]
	public void acceptance_tests_pass(string testCase, Expectation expectation)
	{
		_testOutputHelper.WriteLine(ToYaml(expectation));
		if (expectation.exception == null)
		{
			Ast.Node node = _parser.Parse(expectation.expression);
			node.Should().Be(expectation.expected_ast.ToNode());
		}
		else
		{
			FluentActions.Invoking(() => _parser.Parse(expectation.expression))
				.Should().Throw<CucumberExpressionException>().WithMessage(expectation.exception);
		}
	}

	public class Expectation
	{
		public string expression;
        // ReSharper disable once InconsistentNaming
        public YamlableNode expected_ast;
		public string exception;
	}
	public class YamlableNode
	{
		public Ast.NodeType type;
		public List<YamlableNode> nodes;
		public string token;
		public int start;
		public int end;

		public Ast.Node ToNode()
		{
			if (token != null)
			{
				return new Ast.Node(type, start, end, token);
			}
			else
			{
				return new Ast.Node(type, start, end, nodes.Select(n => n.ToNode()).ToArray());
			}
		}
	}
}
